Add blog post for announcing release of v21 GA (#1872)
vitessio/website: content/en/blog/2024-10-29-announcing-vitess-21.md

---
author: 'Vitess Maintainer Team'
date: 2024-10-29
slug: '2024-10-29-announcing-vitess-21'
tags: [ 'release', 'Vitess', 'v21', 'MySQL', 'kubernetes', 'operator', 'vreplication', 'multi-tenancy', 'Usability', 
        'Online DDL' ]
title: 'Announcing Vitess 21'
description: "Vitess 21 is now Generally Available"
---

# **Announcing Vitess 21**

We're delighted to announce the release of [Vitess 21](https://github.com/vitessio/vitess/releases/tag/v21.0.0) along
with [version 2.14.0](https://github.com/planetscale/vitess-operator/releases/tag/v2.14.0) of the Vitess Kubernetes
Operator.

Version 21 focuses on enhancing query compatibility, improving cluster management, and expanding VReplication
capabilities, with experimental support for atomic distributed transactions and recursive CTEs. Key features include

> reference table materialization, multi-metric throttler support, and enhanced online DDL functionality. Backup and

Reviewer: ```suggestion
reference table materialization, multi-metric throttler support, and enhanced Online DDL functionality. Backup and
```

Author: Done

restore processes benefit from a new **mysqlshell** engine, while **vexplain** now offers detailed execution traces and
schema analysis. The Vitess Kubernetes Operator introduces horizontal auto-scaling for VTGate pods and Kubernetes 1.31
support, improving overall scalability and deployment flexibility.

## What's New in Vitess 21

* **Query Compatibility**: Experimental support for atomic distributed transactions and recursive CTEs
* **VReplication**: Reference table materialization, dynamic workflow configuration
* **Cluster Management And VTOrc**: More metrics in VTOrc to track errant GTIDs
* **Throttler**: multi-metric support
* **Online DDL**: Various improvements
* **Backup & restore**: Experimental mysqlshell engine
* **Vitess Operator**: VTGate scaling, image customization, Kubernetes 1.31 support
* **VTAdmin**: VReplication workflow creation and management, distributed transaction management

> * **VExplain**: **vexplain trace** for detailed query execution insights, **vexplain keys** for analyzing sharding key

Reviewer: Should these be code-formatted instead of bold? `vexplain trace` and `vexplain keys`?

Author: Changed

  usage and optimizing query performance

## Let’s Dive Deeper

Let’s take a deeper look at some key highlights of this release.

### Query Compatibility

#### Atomic Distributed Transactions

We’re reintroducing atomic distributed transactions with a revamped, more resilient design. This feature now offers

> deeper integration with core Vitess components and workflows, such as OnlineDDL and VReplication (including operations

Reviewer 2: ```suggestion
deeper integration with core Vitess components and workflows, such as Online DDL and VReplication (including operations
```

Author: Done

The tablet throttler has been redesigned
with [new multi-metric support](https://vitess.io/docs/21.0/reference/features/tablet-throttler/). With this, the
throttler now handles more than just replication lag or custom queries, but instead can work with multiple metrics at
the same time, and check for different metrics for different clients or for different workflows. This gives users better
control over the throttler allowing them to fine-tune its behavior based on their specific production requirements.

Several new metrics have been introduced in v21, with plans to expand the list of available metrics in later versions.

The multi-metric throttler in v21 is backward compatible with the v20 throttler. It is possible to have a v20 primary
tablet collecting throttler data from a v21 replica tablet, and vice versa. This backward compatibility will be removed
in v22, where all tablet throttlers will be expected to communicate multi-metric data.

Other key throttler changes:

* With the above, the sub-flags `--check-as-check-self` and `--check-as-check-shard` to
  the `UpdateThrottlerConfig` command are deprecated and slated to be removed in a future version. \

>   Similarly, `SHOW VITESS_THROTTLER STATUS` and `SHOW VITESS_THROTTLED_APPS` queries, and all `/throttler/` 

Reviewer: ```suggestion
* Similarly, `SHOW VITESS_THROTTLER STATUS` and `SHOW VITESS_THROTTLED_APPS` queries, and all `/throttler/` 
```

Author: Done

>   Similarly, ``SHOW VITESS_THROTTLER STATUS`` and ``SHOW VITESS_THROTTLED_APPS`` queries, and all ``/throttler/`` 
> * API access points (with the exception of ``/throttler/check``) are deprecated and slated to be removed in v22.

Reviewer 2: ```suggestion
  Similarly, ``SHOW VITESS_THROTTLER STATUS`` and ``SHOW VITESS_THROTTLED_APPS`` queries, and all ``/throttler/`` API access points (with the exception of ``/throttler/check``) are deprecated and slated to be removed in v22.
```

Reviewer 2: Also, why are there double backticks everywhere? Must be a copy+paste issue from Google Docs.

Author: Thanks. Fixed the list item formatting issue. Also replaced double backticks with single: the Google Doc extension had generated the double ticks and I assumed there was some reason for it since it seemed to look fine in the preview.
